Integrated knife-opc into Chef

snehaldwivedi opened the pull request:
Signed-off-by: snehaldwivedi <sdwivedi@msystechnologies.com>

<!--- Provide a short summary of your changes in the Title above -->

## Description

- Merge this knife plugin into Chef.
- Moved `user password` command to core chef

## Related Issue
https://github.com/chef/knife-opc/issues/48
https://github.com/chef/knife-opc/issues/57
https://github.com/chef/chef/issues/3517
https://github.com/chef/chef/issues/3514 (covers `knife user create` should check file permissions), https://github.com/chef/chef/issues/3515, https://github.com/chef/chef/issues/3516

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I have read the **CONTRIBUTING** document.
- [x] I have run the pre-merge tests locally and they pass.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

stevendanna commented on lib/chef/knife/opc_org_create.rb:
```diff
@@ -0,0 +1,67 @@
+module Opc
+  class OpcOrgCreate < Chef::Knife
+    category "CHEF ORGANIZATION MANAGEMENT"
+    banner "knife opc org create ORG_SHORT_NAME ORG_FULL_NAME (options)"
```
We should consider removing the `opc` prefix from all of these commands.

tas50 replied:
Agreed

stevendanna commented on lib/chef/knife/opc_org_list.rb:
```diff
@@ -0,0 +1,45 @@
+    def run
+      results = root_rest.get("organizations")
+      unless config[:all_orgs]
+        results = results.select { |k, v| !(k.length == 20 && k =~ /^[a-z]+$/) }
```
I _think_ this filter is from when we used to pre-generate organizations because organization took a long time. I believe this is no longer the default so perhaps it would be best to turn this filtering off by default and it can end up filtering out legit organizations.

stevendanna commented on lib/chef/org/group_operations.rb:
```diff
@@ -0,0 +1,60 @@
+
+class Chef
+  class Org
+    module GroupOperations
```
I wonder if we really need a separate module for this.  It was structured this way in the plugin because we lived outside the Chef codebase, but we could probably add these features directly to the Chef::Org class or create a new Chef::Group class that would abstract away some of these features.

stevendanna commented on lib/chef/knife/opc_user_create.rb:
```diff
@@ -0,0 +1,101 @@
+module Opc
+  class OpcUserCreate < Chef::Knife
+    category "CHEF ORGANIZATION MANAGEMENT"
+    banner "knife opc user create USERNAME FIRST_NAME [MIDDLE_NAME] LAST_NAME EMAIL PASSWORD"
```
We already have a set of `knife user` commands that I think should cover most of the user-related commands in this PR.

tas50 replied:
I'd like to see one or the other, but not both user commands. Since every chef server is a server with an org at this point we should just make any missing functionality part of the existing `knife user` operations.

stevendanna commented on lib/chef/mixin/root_rest.rb:
```diff
@@ -0,0 +1,31 @@
+        # Rather than upgrade all of this code to move to v1, the goal is to remove the
+        # need for this plugin.  See
+        # https://github.com/chef/chef/issues/3517
+        @root_rest ||= Chef::ServerAPI.new(Chef::Config[:chef_server_root], { api_version: "0" })
```
I believe that the existing user plugins were converted to v1 and that none of the other APIs here really care about the API version so we can _probably_ don't need to specify the api_version if we end up removing the user-related commands from this PR.

lamont-granquist replied:
Yeah this should go away and we should not be using v0 here.

snehaldwivedi commented:
NOTE: this test is failing due to changes in gem chef/.bundle/ruby/2.7.0/gems/tty-prompt-0.21.0/lib/tty/prompt/question.rb
 `attr_accessor :echo`  need to be added here

snehaldwivedi commented:
I am working on all review comments also I found that both `knife user` and `knife opc user` commands don't match completely, like for example `knife opc user list` also provide list of user associated with organisations which is not available in `knife user list ` command. So I think we need to merge that too and started working on that too.

lamont-granquist commented on lib/chef/knife/org_delete.rb:
```diff
@@ -0,0 +1,35 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+require_relative "../mixin/root_rest"
```
this mixin still needs to go away

tas50 commented on lib/chef/knife/user_list.rb:
```diff
@@ -1,6 +1,6 @@
 #
 # Author:: Steven Danna (<steve@chef.io>)
-# Copyright:: Copyright (c) Chef Software Inc.
+# Copyright:: Copyright 2011-2016 Chef Software, Inc.
```
None of these copyrights need to get updated

vsingh-msys commented on lib/chef/knife/user_create.rb:
```diff
@@ -88,20 +101,48 @@ def run
       end
+
+      def prompt_for_password
+        ui.ask("Please enter the user's password: ") { |q| q.echo = false }
```
```suggestion
        ui.ask("Please enter the user's password: ", echo: false)
```

btm commented:
@snehaldwivedi if `knife org edit` survives into core, please test it to see that it opens the editor? https://github.com/chef/knife-opc/issues/57

lamont-granquist commented on lib/chef/knife/org_show.rb:
```diff
@@ -15,19 +15,21 @@
         org_name = @name_args[0]
-        ui.output root_rest.get("organizations/#{org_name}")
+        org = Chef::Org.from_hash({ "name" => org_name })
+        ui.output org.chef_rest.get("organizations/#{org_name}")
```
This seems like this should be pushed into the Chef::Org object.

I haven't looked but we should be able to do something to get a Chef::Org object for the org_name directly rather than going in through `from_hash` and then the `get` call should be hidden from the caller.   That could be combined into one static method if necessary like `Chef::Org.load(org_name)` (unless that collides with an existing method name but something like that)

snehaldwivedi replied:
okay I will look into that

vsingh-msys replied:
@lamont-granquist I would prefer to add `root_rest` method to `knife.rb` similar to `rest` so that it will be available to all subclasses as we have more root level operations into chef core.

```
    def root_rest
      @root_rest ||= begin
        require_relative "server_api"
        Chef::ServerAPI.new(Chef::Config[:chef_server_root])
      end
    end
```

lamont-granquist replied:
So the point here though is that AFAIK we should not be using protocol v0, and I don't understand why this PR is spreading it everywhere.

The chef/user.rb that defines the chef_rest_v0 object has a large banner at the top of it that Chef::User should be replaced by Chef::UserV1 in Chef 13 (that didn't happen) but AFAIK it should all go away.  I have no idea why knife-opc was using v0, that may have been due to some kind of backwards compatibility with now-ancient chef-servers.  Without knowing what breaks, my position is that it should be removed and we should be using our normal v2 rest objects consistently.

stevendanna replied:
knife-opc was using v0 simply because we needed it to work ASAP and that was the smallest possible change, so :+1: to doing it right this time around.

tas50 commented on lib/chef/knife/user_delete.rb:
```diff
@@ -23,21 +23,133 @@ class Knife
+      option :no_disassociate_user,
+        long: "--no-disassociate-user",
+        short: "-d",
+        description: "Don't disassociate the user first"
```
disassociate from what? This description isn't very helpful

snehaldwivedi replied:
I will update it

snehaldwivedi replied:
for disassociate user I will update once this is merge https://github.com/chef/knife-opc/pull/36 as it has changes related to it

snehaldwivedi replied:
@tas50 can you please confirm if https://github.com/chef/knife-opc/pull/56 is good to add changes in this PR?

lamont-granquist commented on lib/chef/userable.rb:
```diff
@@ -0,0 +1,106 @@
+
+# common methods of User and UserV1
+class Chef
+  class Userable
```
This should probably be something like Chef::Mixin::UserBase or something, although I don't like the Mixin directory naming scheme, but I can't think of anything better offhand right now.

lamont-granquist replied:
@phiggins might have some better ideas than i do

phiggins replied:
I agree, I'd prefer something like `AbstractUser` or `BaseUser` to `Userable`.

lamont-granquist commented on lib/chef/userable.rb:
```diff
@@ -0,0 +1,106 @@
+
+    def chef_root_rest_v1
+      @chef_root_rest_v22 ||= Chef::ServerAPI.new(Chef::Config[:chef_server_root], { api_version: "1" })
+    end
```
We shouldn't need explicit v1 here, the Userv1 API is compatible with the v2 API as well and we should be able to use the same rest object that all the other APIs consume, and not have to recreate a new one.

The naming of the instance variable is also a bit odd

phiggins replied:
For memoization, the ruby convention is to name the instance variable the same as the method. It should definitely not get reused between methods.

tas50 commented on lib/chef/knife/user_create.rb:
```diff
@@ -27,7 +26,7 @@ class UserCreate < Knife
       attr_accessor :user_field
 
       deps do
-        require_relative "../user_v1"
```
We can safely assume v1 support as the minimum API version at this point so nothing here should be moving back to the "user" class. Everything should just code to the _v1 APIs.

snehaldwivedi commented:
Regarding https://chefio.slack.com/archives/G08912B8E/p1600190312027500. For explanation,I have used used V2 tested all the commands and they are working fine. I have created an common class `Userable` and moved common codes of  `user.rb` & `userv1.rb` into it, which was required in my commands and used that like this https://github.com/chef/chef/pull/10187/files#diff-b6a20c4a70b8b7a50778c9747fd78c9eL38 
https://github.com/chef/chef/pull/10187/files#diff-ebda7b27eff28ff699ce69fe7ba6e8b1R21. So this PR is currently using v2 objects.

Now currently, I working on naming convention comments and will update the PR once completed with that.

lamont-granquist commented:
This still needs:

- remove the userable mixin, that has nothing to do with this PR and it is complicating the issues

- write the plugin using the userv1 object with the standard knife rest object.

- don't change the api of the existing knife plugins

lamont-granquist commented:
So the knife commands should all just use this rest object:

https://github.com/chef/chef/blob/master/lib/chef/knife.rb#L640-L645

All the knife commands inherit that for free, and that can be injected into the constructor.

Inside of chef-client everything should be getting this rest object injected into the constructor:

https://github.com/chef/chef/blob/master/lib/chef/run_context.rb#L50-L55

Most of the objects have a signature like this and allow injecting the rest client:

    def initialize(chef_server_rest: nil, logger: nil)

It looks like client_v1 does not do that, but that is fine, just use the client_v1 object and please ignore the issue.  There needs to be some thought around cleaning that up.  Most likely what needs to happen is NOT any cleanup around the rest objects directly but that the users_v1 object needs to be converted to use ./mixin/versioned_api.rb like the cookbook_manifest.rb object does (users_v1 precedes the creation of the versioned_api mixin and was never converted), but that is WELL outside of the scope of this PR.

snehaldwivedi commented:
I have removed `abstract_user` as we are not covering cleanup #10458 on this PR. Updated API version back to 1 and used `user_v1` for commands that require any helper methods else used directly  `rest` objects.

snehaldwivedi commented:
@tas50 Can you please have a look into the PR for merge or if any more changes required form my side?

lamont-granquist commented:
You should be able to eliminate the `@root_rest` API object entirely and just use the default knife rest object.  That is a v2 REST object but it should work fine with the users_v1 object.  They don't have to match the rest object just needs to be a higher version, which it should be.  It would break with users_v0 which is deprecated at this point.

snehaldwivedi commented:
Hi @lamont-granquist 

As per my analysis, I cannot use the default `rest` object as its uses `chef_server_url` which will not provide us the exact result required or the opc commands. Because `chef_server_url` is URL for the specific organizations which will not provide the proper result like if we need a list of all organization, global user requirement in this [ticket](https://github.com/chef/chef/issues/3513).

If we can change `chef_server_url` to `chef_server_root` in [rest]( https://github.com/chef/chef/blob/add5264d5f7f727c0b1c5a265f91468cfa0e7cb6/lib/chef/knife.rb#L643 ) object then we can use it. Please let me know if you have any suggestions on this or any other way to do this?

lamont-granquist commented:
Okay the communication failure is now a bit more obvious.  You can do that, but you need to remove `{ api_version: "1" }` off of the arguments, that's the problem that I'm having with it.

lamont-granquist commented on lib/chef/knife.rb:
```diff
@@ -640,7 +640,7 @@ def test_mandatory_field(field, fieldname)
       @rest ||= begin
         require_relative "server_api"
-        Chef::ServerAPI.new(Chef::Config[:chef_server_url])
+        Chef::ServerAPI.new(Chef::Config[:chef_server_root])
```
doesn't this change the rest API for all the rest of the knife commands though?

i was okay with the root_rest object but just didn't want it v1.
